File: 49.Group-Anagrams/map.py
class Solution:
    def groupAnagrams(self, strs: list[str]) -> list[list[str]]:
        n = len(strs)
        groups = {}

        # O(n * klogk)
        for i in range(n):
            key = tuple(sorted(strs[i]))
            anagrams = groups.get(key, [])
            anagrams.append(strs[i])
            groups[key] = anagrams

        # O(n)
        return list(groups.values())

File: 49.Group-Anagrams/map.rs
use std::collections::HashMap;

impl Solution {
    pub fn group_anagrams(strs: Vec<String>) -> Vec<Vec<String>> {
        let mut words: HashMap<_, Vec<String>> = HashMap::new();

        // O(n * klogk)
        for word in strs {
            let mut arr: Vec<_> = word.chars().collect();
            arr.sort();

            words.entry(arr).or_default().push(word);
        }

        // O(n)
        words.into_values().collect()
    }
}
